walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,105 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Set;
 
 public final class FindMeetingQuery {
+  /**
+   * This algorithm creates a List of all the times at least one attendee is busy, then uses that
+   * List to find all the times that everyone is available The worst-case runtime is O(n^2)
+   */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    Collection<String> attendeesRequested = request.getAttendees();
+    int meetingDuration = (int) request.getDuration();
+
+    // check for no attendees
+    if (attendeesRequested.isEmpty()) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+
+    // check duration of meeting request
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+      return Arrays.asList();
+    }
+
+    List<TimeRange> badTimes = getBadTimes(events, attendeesRequested);
+
+    // list to hold all the available times (everything not in badTimes)
+    List<TimeRange> goodTimes = getGoodTimes(badTimes, meetingDuration);
+    return goodTimes;
+  }
+
+  /** Returns a List with all the times that requested attendees have conflicts. */
+  private List<TimeRange> getBadTimes(
+      Collection<Event> events, Collection<String> attendeesRequested) {
+    List<TimeRange> badTimes = new ArrayList<TimeRange>();
+
+    // iterate through all events and add  time to badTimes if attendee is in attendeesRequested
+    for (Event conflictEvent : events) {
+      Set<String> busyAttendees = conflictEvent.getAttendees();
+
+      // loop to see if any attendees in conflictEvent are in attendeesRequested
+      Boolean conflict = false;
+      Iterator<String> eventIter = busyAttendees.iterator();
```

Comment:
nit: maybe name this attendeeIter to make it clear what the value represents.

optional: this while loop can also be done using a for loop and a break statement if you prefer that approach.
```
  for (String attendee : busyAttendees) {
    if (attendeesRequests.contains(attendee) {
      badTimes.add(conflictEvent.getWhen());
      break;
    }
  }
```

Reply:
I agree, that code is much easier to follow - thanks! I was hesitant to use the break statement because one of my professors said it was bad style, but I think it makes sense here :)